tests/prague/eip2537_bls_12_381_precompiles/helpers.py

```diff
@@ -61,3 +65,579 @@ def vectors_from_file(filename: str) -> List:
+                    else:
+                        point2 = PointG2(
+                            (int(x.coeffs[0]), int(x.coeffs[1])),
+                            (Spec.P - int(y.coeffs[0]), Spec.P - int(y.coeffs[1])),
+                        )
+
+                    # Verify points have the required properties
+                    point1_on_curve = cls.is_on_curve_g2(point1.x, point1.y)
+                    point2_on_curve = cls.is_on_curve_g2(point2.x, point2.y)
+                    point1_in_subgroup = cls.check_in_g2_subgroup(point1)
+                    point2_in_subgroup = cls.check_in_g2_subgroup(point2)
+
+                    # Return required point if found based on properties
+                    if on_curve == point1_on_curve and in_subgroup == point1_in_subgroup:
+                        return point1
+                    if on_curve == point2_on_curve and in_subgroup == point2_in_subgroup:
+                        return point2
+
+                except Exception:
+                    continue
+
+        raise ValueError(
+            (
+                f"Failed to find G2 point by x={x_value},",
+                "in_subgroup={in_subgroup},",
+                "on_curve={on_curve}",
+            )
+        )
+
+    # G1 points by x coordinate (near or on the x value)
+    @classmethod
+    def generate_g1_point_in_subgroup_by_x(cls, x_value: int) -> PointG1:
+        """G1 point that is in the r-order subgroup with x-coordinate by/on the given value."""
+        return cls.find_g1_point_by_x(x_value, in_subgroup=True, on_curve=True)
+
+    @classmethod
+    def generate_g1_point_not_in_subgroup_by_x(cls, x_value: int) -> PointG1:
+        """G1 point that is NOT in the r-order subgroup with x-coordinate by/on the given value."""
+        return cls.find_g1_point_by_x(x_value, in_subgroup=False, on_curve=True)
+
+    @classmethod
+    def generate_g1_point_not_on_curve_by_x(cls, x_value: int) -> PointG1:
+        """G1 point that is NOT on the curve with x-coordinate by/on the given value."""
+        return cls.find_g1_point_by_x(x_value, in_subgroup=False, on_curve=False)
+
+    @classmethod
+    def generate_g1_point_on_isomorphic_curve_by_x(cls, x_value: int) -> PointG1:
+        """
+        G1 point that is on an isomorphic curve (not standard curve)
+        but in the r-order subgroup with x-coordinate by/on the given value.
+
+        Uses cofactor multiplication to ensure the point is in the correct subgroup.
+        """
+        return cls.find_g1_point_by_x(x_value, in_subgroup=True, on_curve=False)
+
+    # G1 random points - all methods updated with seed parameter
+    @classmethod
+    def generate_random_g1_point_in_subgroup(cls, seed=None) -> PointG1:
+        """Generate a random G1 point that is in the r-order subgroup."""
+        seed_value = cls.BASE_SEED if seed is None else cls.BASE_SEED + seed * 10
+        random.seed(seed_value)
+        x = random.randrange(Spec.P)  # Random field element in Fp
```

Comment:
I would lean towards making `seed` obligatory and using a hash function instead.

I can see that `Spec.P` is 380 bits long, so we could perhaps use sha384? (I think it should be in the hashlib library).